### PR TITLE
Add support for distributivity refactoring of + and *.

### PR DESCRIPTION
**Pull request by martin-cs**
This helps an EDA customer.

### CODE REVIEW

**Comment by martin-cs** [comment]
I've fixed the compile errors; can you have another look at this?

**Comment by martin-cs** [comment]
Candidate for the next release?

File: src/util/simplify_expr_int.cpp [hunk 1]
```diff
@@ -504,6 +505,133 @@ bool simplify_exprt::simplify_plus(exprt &expr)
       }
     }
 
+
+    // Distributivity refactoring
+    typedef hash_set_cont<exprt, irep_hash> occurence_sett;
+    typedef hash_map_cont<exprt, occurence_sett, irep_hash> factorst;
+    typedef hash_map_cont<exprt, unsigned, irep_hash> occurence_countt;
```
Comment (tautschnig) [inline src/util/simplify_expr_int.cpp]: It's std::unordered_map now

File: src/util/simplify_expr_int.cpp [hunk 2]
```diff
@@ -392,6 +392,7 @@ bool simplify_exprt::simplify_plus(exprt &expr)
     return true;
 
   bool result=true;
+  bool recurse_needed = false;
```
Comment (tautschnig) [inline src/util/simplify_expr_int.cpp]: Looks like this patch needs a run of the linter.

File: src/util/simplify_expr_int.cpp [hunk 3]
```diff
@@ -504,6 +505,133 @@ bool simplify_exprt::simplify_plus(exprt &expr)
...
+	fac[*it].insert(*it);
+	++counts[*it];
+      }
+    }
+
+    // heuristic : factor out the most common multiplicand ...
+    occurence_countt::const_iterator max_occ = counts.begin();
+    for (occurence_countt::const_iterator i = counts.begin();
+	 i != counts.end();
+	 ++i)
```
Comment (tautschnig) [inline src/util/simplify_expr_int.cpp]: ranged for

File: src/util/simplify_expr_int.cpp [hunk 4]
```diff
@@ -504,6 +505,133 @@ bool simplify_exprt::simplify_plus(exprt &expr)
...
+      factored.reserve(max_occ->second);
+      
+      Forall_expr(it, operands)
+      {
+	if (fold->second.find(*it) != fold->second.end())
+	{
+	  // Need to remove one instance of the folded term
+	  // and add to the factored list
+	  exprt tmp(*it);
+	  *it = gen_zero(it->type());
```
Comment (tautschnig) [inline src/util/simplify_expr_int.cpp]: from_integer(0, it->type())

File: src/util/simplify_expr_int.cpp [hunk 5]
```diff
@@ -504,6 +505,133 @@ bool simplify_exprt::simplify_plus(exprt &expr)
...
+	    
+	    bool found = false;
+	    
+	    Forall_operands(jt, tmp)
+	    {
+	      if (*jt == fold->first)
+	      {
+		found = true;
+		//tmp.operands().erase(jt); // Safe as jt not used again.
+		*jt = gen_one(jt->type()); // Erase doesn't catch the unit case
```
Comment (tautschnig) [inline src/util/simplify_expr_int.cpp]: from_integer(1, jt->type())

**Comment by martin-cs** [comment]
Have applied all of @tautschnig 's comments, linted, rebased and squashed.  Should be good to go.

File: src/util/simplify_expr_int.cpp [hunk 6]
```diff
@@ -511,6 +514,137 @@ bool simplify_exprt::simplify_plus(exprt &expr)
...
+      if(it->id()==ID_mult)
+      {
+        occurence_sett unique_factors; // x*x only adds 1 to x's count
+        forall_operands(jt, *it)
+        {
+          fac[*jt].insert(*it);
+          unique_factors.insert(*jt);
+        }
+
+        for(auto ui : unique_factors)
```
Comment (tautschnig) [inline src/util/simplify_expr_int.cpp]: `const auto &ui : unique_factors`

File: src/util/simplify_expr_int.cpp [hunk 7]
```diff
@@ -511,6 +514,137 @@ bool simplify_exprt::simplify_plus(exprt &expr)
...
+        fac[*it].insert(*it);
+        ++counts[*it];
+      }
+    }
+
+    // heuristic : factor out the most common multiplicand ...
+    assert(!counts.empty());
+    occurence_countt::const_iterator max_occ = counts.begin();
+    for(occurence_countt::const_iterator i = counts.begin();
+        i != counts.end();
```
Comment (tautschnig) [inline src/util/simplify_expr_int.cpp]: In all three lines above: no whitespace around =, !=

File: src/util/simplify_expr_int.cpp [hunk 8]
```diff
@@ -511,6 +514,137 @@ bool simplify_exprt::simplify_plus(exprt &expr)
...
+        ++i)
```
Comment (tautschnig) [inline src/util/simplify_expr_int.cpp]: Could actually be done via ranged for

File: src/util/simplify_expr_int.cpp [hunk 9]
```diff
@@ -511,6 +514,137 @@ bool simplify_exprt::simplify_plus(exprt &expr)
...
+    {
+      if(i->second > max_occ->second &&
+          !i->first.is_constant())
+        max_occ = i;
+    }
+
+    factorst::iterator fold = fac.find(max_occ->first);
```
Comment (tautschnig) [inline src/util/simplify_expr_int.cpp]: I believe that could be a const_iterator

File: src/util/simplify_expr_int.cpp [hunk 10]
```diff
@@ -511,6 +514,137 @@ bool simplify_exprt::simplify_plus(exprt &expr)
...
+
+      plus_exprt bracket(factored[0], factored[1]);
+
+      exprt::operandst::const_iterator it = factored.begin();
+      ++it; ++it;
+      while(it != factored.end())
+      {
+        bracket.operands().push_back(*it);
+        ++it;
+      }
```
Comment (tautschnig) [inline src/util/simplify_expr_int.cpp]: This code really places all of factored into the operands of bracket; you might just do
```
plus_exprt bracket(factored[0], factored[1]);
bracket.operands().swap(factored);
```

**Comment by thomasspriggs** [comment]
I am going to close this PR as it appears that work on it has been inactive for an extended period of time. This is not any judgement on how worthwhile this work is. This is part of an effort to reduce the number of open PRs which are not being actively worked on. If you still think it would be worthwhile to get this merged then the branch should be re-based on the latest version of develop and the PR re-opened.

**Pull request closed without merging**